Clean up some old DynamicSelfType debt now that we have an implementation of SE-0068

slavapestov opened the pull request:
Builds on https://github.com/apple/swift/pull/23991, cleaning up some older code that was making the full implementation of SE-0068 difficult.

- Support immutable (computed) properties and subscripts with `Self` type (this was @johnno1962's PR).
- Stop wrapping Self-returning protocol methods with DynamicSelfType.
- Remove `FuncDecl::hasDynamicSelf()` and the weird TypeRepr inspection that was used to compute it.

Fixes #52726.

johnno1962 committed:
Story so far

johnno1962 committed:
Some progress

johnno1962 committed:
Make properties and subscripts returning Self overridable.

johnno1962 committed:
Seems to work

johnno1962 committed:
Remove debug messages

johnno1962 committed:
Tidy-up

johnno1962 committed:
Revert some tidy-ups.

johnno1962 committed:
CovariantReturnConversionExpr required.

johnno1962 committed:
Resolve failing constraint for subscripts returning Self

johnno1962 committed:
Move validation of DynamicSelf to Decl checks.

johnno1962 committed:
Resolve conflict

johnno1962 committed:
Handle Optionals

johnno1962 committed:
Update tests

johnno1962 committed:
Dynamic `self` in properties and subscripts

johnno1962 committed:
A couple of comments

johnno1962 committed:
More additional tests

slavapestov committed:
Fix compile error and test failure on latest master

slavapestov committed:
Sema: Remove 'lookupDC'

slavapestov committed:
Sema: Stop wrapping Self returns in protocols with DynamicSelfType

slavapestov committed:
AST: Fix Type::subst() edge case with DynamicSelfType

slavapestov committed:
Sema: Small cleanup for simplifyTypeImpl()

slavapestov committed:
Sema: Tidy up support for DynamicSelfType return type in @objc methods

slavapestov commented:
@johnno1962 Here is what I have so far. The first 16 patches are from your PR, rebased onto latest master. The next patch fixes a compiler error that snuck in along the way.

The rest are what I have working so far in cleaning up some older code. I'm trying to get rid of `TypeResolverContext::DynamicSelfResult`, `FuncDecl::hasDynamicSelf()`, and some other stuff. There's not much to look at yet, but when it's done it will allow us to simplify some of the new Self-related diagnostic checks.

johnno1962 commented:
LGTM as they say. I see you're taking the opportunity to tidy a few things up. You're braver than me!

slavapestov committed:
Sema: Remove TypeResolverContext::DynamicSelfResult

slavapestov committed:
AST: Add AbstractFunctionDecl::hasDynamicSelfResult() and use it

This calculates a result directly from the function's result type
instead of checking a bit that was previously set by the type
checker. Also, always returns true for constructors to simplify
some callers.

slavapestov committed:
AST: Remove FuncDecl::hasDynamicSelf()

slavapestov committed:
Sema: Merge two places where we resolve 'Self'

slavapestov committed:
Sema: TypeChecker::conformsToProtocol() should only compute conditional requirements if checking them

This fixes a crash with the next commit.

slavapestov committed:
Sema: Consolidate diagnostics for invalid usage of DynamicSelfType

slavapestov committed:
Sema: Stored properties cannot have DynamicSelfType

slavapestov committed:
ClangImporter: Don't wrap protocol Self return in DynamicSelfType

slavapestov committed:
AST: Address a FIXME in adjustSuperclassMemberDeclType()

This is NFC for now, but fixes tests after we remove eraseDynamicSelfType().

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@johnno1962 I went back and forth several times on whether to use the approach from your PR of allowing Self almost anywhere and then walking decl types after the fact to check invariants, over the old way of only allowing Self in certain cases. What I settled on is a hybrid approach.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
Source compat failure is an unrelated UPASS.

johnno1962 commented:
Thanks! Great to see this tidied up properly. And now... my next crusade [inheriting protocol extensions](https://github.com/apple/swift/pull/25786). Any thoughts you have appreciated.

slavapestov commented:
@johnno1962 That one is above my level of ability. I know there's an entry in the generics manifesto about it. The tricky part is that adding new protocols to the hierarchy like that changes generic signature minimization and witness table layout; we need to figure out how to allow that without breaking resilience. A forum post tagging @DougGregor (or e-mail him directly :)  ) is your best bet.

Maybe start with a pitch on the forums so that folks can discuss the language design issues and work out the corner cases first?
